src/main/java/com/streamr/client/dataunion/DataUnion.java

```diff
@@ -0,0 +1,166 @@
+package com.streamr.client.dataunion;
+
+import com.streamr.client.dataunion.contracts.DataUnionMainnet;
+import com.streamr.client.dataunion.contracts.DataUnionSidechain;
+import com.streamr.client.dataunion.contracts.IERC20;
+import com.streamr.client.utils.Web3jUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.web3j.abi.TypeEncoder;
+import org.web3j.abi.datatypes.*;
+import org.web3j.abi.datatypes.generated.Uint256;
+import org.web3j.crypto.Credentials;
+import org.web3j.crypto.ECKeyPair;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.methods.response.TransactionReceipt;
+import org.web3j.tx.Contract;
+import org.web3j.tx.ManagedTransaction;
+import org.web3j.utils.Numeric;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.math.BigInteger;
+import java.util.ArrayList;
+
+import static com.streamr.client.dataunion.DataUnionClient.signWithdraw;
+import static com.streamr.client.utils.Web3jUtils.waitForCodeAtAddress;
+import static com.streamr.client.utils.Web3jUtils.waitForCondition;
+
+public class DataUnion {
+    private static final Logger log = LoggerFactory.getLogger(DataUnion.class);
+    protected static final long STATUS_NONE = 0;
+    protected static final long STATUS_ACTIVE = 1;
+    protected static final long STATUS_INACTIVE = 2;
+
+    // Contract.web3j is protected, so need to access with reflection. more bad web3j design!
+    private static Web3j getWeb3j(Contract c) {
+        try {
+            return (Web3j) FieldUtils.readField(c, "web3j", true);
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            throw new RuntimeException(e);
+        }
```

Comment:
We should either handle error locally or throw it, but not both.

Reply:
well this is an exotic error that occurs if your JVM doesn't support field access modification. This code is ugly, but no other way to get Web3j object from Contract. There's lots of crap like this in Web3j lib. They hide too much.

Reply:
actually... I can remove this method and pass the connector. the constructor is protected, so doesn't expose anything

Reply:
My point is about general error handling. While this is not the most relevant case since method is private but similar pattern can be found from EstimatedGasProvider.java:51-52 where method is public. Error is logged here which should be enough of error handling for one error, but error is thrown after logging.
We should either:
- Log the error and throw nothing
or
- Throw the error and log nothing
or
- Handle error in a relevant way and throw/log nothing

Logging and throwing is duplicated error handling and leads to logging the same error several times.